src/php/Lang/Set.php

```diff
@@ -0,0 +1,163 @@
+<?php
+
+namespace Phel\Lang;
+
+use ArrayAccess;
+use Countable;
+use InvalidArgumentException;
+use Iterator;
+use Phel\Printer;
+
+class Set extends Phel implements Countable, Iterator, ICons, IPush, IConcat
+{
+
+    /**
+     * @var mixed[]
+     */
+    protected $data = [];
+
+    /**
+     * Constructor
+     *
+     * @param mixed[] $data A list of all values
+     */
+    public function __construct(array $data)
+    {
+        $this->setData($data);
+    }
+
+    /**
+     * Create a new set.
+     *
+     * @param mixed[] $values A list of all values
+     *
+     * @return Set
+     */
+    public static function create(...$values): Set
+    {
+        return new Set($values);
+    }
+
+    public function isTruthy(): bool
+    {
+        return true;
+    }
+
+    public function hash(): string
+    {
+        return spl_object_hash($this);
+    }
+
+    public function count(): int
+    {
+        return count($this->data);
+    }
+
+    public function current()
+    {
+        return current($this->data);
+    }
+
+    public function next()
+    {
+        next($this->data);
+    }
+
+    public function key()
+    {
+        return key($this->data);
+    }
+
+    public function valid()
+    {
+        return key($this->data) !== null;
+    }
+
+    public function rewind()
+    {
+        reset($this->data);
+    }
+
+    public function cons($x): ICons
+    {
+        array_unshift($this->data, $x);
+        return $this;
+    }
+
+    public function push($x): IPush
+    {
+        $hash = $this->offsetHash($x);
+        $this->data[$hash] = $x; // Don't need to check if $x is already there, just override.
+
+        return $this;
+    }
+
+    public function concat($xs): IConcat
+    {
+        foreach ($xs as $x) {
+            $this->push($x);
+        }
+
+        return $this;
+    }
+
+    public function union(Set $set): IConcat
+    {
+        return $this->concat($set->toPhpArray());
+    }
+
+    public function intersection(Set $set): IConcat
+    {
+        return $this->applySet($set, 'array_intersect');
+    }
+
+    public function difference(Set $set): IConcat
+    {
+        $rightDiff = new Set($set->toPhpArray());
```

Comment:
I don't really understand what you are doing here.

I would just go for:
```
$this->data = array_diff_key($this->data, $set->toPhpArray());
```

Reply:
Indeed, `array_diff_key` is definitely a better approach, I updated my code to use this function. But the biggest weirdness in my code comes from the way PHP implements `array_diff`.

Take this code:
```php
$foo = [1, 2];
$bar = [2, 3];
$foobar = array_diff($foo, $bar);
```
Here `$foobar` will not equals to `[1, 3]` as expected, but to `[1]`. That's why I have to execute  array_diff two times, to get the diff from both arrays, and then merge them. The same is true for `array_diff_key`.

But if you know a better solution, I would be happy to know it, because it's actually not very pretty neither readable.